### PR TITLE
Read Lakehouse Data in Batches for Extracting Time Series Data from Fabric

### PR DESCRIPTION
**Pull request by muralidharm**
This PR introduces changes that enable the CDF Fabric Extractor to read data in batches rather than loading the whole Lakehouse table at once.  Batch sizes are configurable, and the data is filtered based on the latest timestamp read per external ID.
## Code Changes
- Configuration - New config value `read_batch_size` under `source` that sets the size of the data batch to be retrieved from the Lakehouse
- `convert_lakehouse_data_to_df_batch` - A generator method that yields batches of data based on the configured batch size.  This method is meant to be called per external id and sets a filter condition based on the external id and an optional latest timestamp.
- Helper methods
   - `retrieve_external_ids_from_lakehouse` - Uses (pyarrow)[https://arrow.apache.org/docs/python/index.html] to retrieve all unique external id's from the Lakehouse data efficiently.
   - `get_timeseries_latest_timestamps` - Based on the list of external id's, retrieves their state from the state store if it exists.  Returns a dictionary of external id's and timestamps.
- Refactor - `convert_lakehouse_data_to_df_batch` and `write_timeseries_data_to_cdf` were put into a new method `extract_time_series_data` to make the extractor more testable
- Unit tests - Unit tests added for the new functions and existing unit tests updated
- Integration Tests - Existing integration test `test_extractor_timeseries_service` was updated to test cases with different batch sizes.  A new integration test was added to test that the timestamp filter works properly

## Additional Notes
Follow up work is needed to apply the same pattern to reading event data and raw data from Lakehouse.

### CODE REVIEW

File: cdf_fabric_replicator/extractor.py [hunk 1]
```diff
@@ -317,6 +345,31 @@ def convert_lakehouse_data_to_df(self, file_path: str, token: str) -> DataFrame:
...
             raise e
 
+    def convert_lakehouse_data_to_df_batch(
+        self, file_path: str, external_id: str, timestamp: str, token: str
+    ) -> Generator[DataFrame, None, None]:
+        try:
+            dt = DeltaTable(
+                file_path,
+                storage_options={"bearer_token": token, "user_fabric_endpoint": "true"},
+            )
```
Comment (xtophs) [inline cdf_fabric_replicator/extractor.py:355]: should this be use_fabric_endpoint? alternatively, did you look into the whole range of azure storage options (https://docs.rs/object_store/latest/object_store/azure/enum.AzureConfigKey.html#variants)?

File: cdf_fabric_replicator/config.py [hunk 2]
```diff
@@ -45,6 +45,7 @@ class SourceConfig:
     data_set_id: str
     event_path: Optional[str] = None
     raw_time_series_path: Optional[str] = None
+    read_batch_size: int = 1000
```
Comment (xtophs) [inline cdf_fabric_replicator/config.py:48]: Great to see the batch size configurable. it doesn't look like the batch size is documented in the README.

File: tests/unit/test_extractor.py [hunk 3]
```diff
@@ -415,6 +405,33 @@ def test_upload_files_from_abfss_cognite_error(
     test_extractor.logger.error.assert_called_once()
 
 
+def test_retrieve_external_ids_from_lakehouse(test_extractor, mocker):
+    # Mock DeltaTable to return a pyarrow table with the externalId column
+    df = pd.DataFrame({"externalId": ["id1", "id2", "id1", "id1"]})
```
Comment (xtophs) [inline tests/unit/test_extractor.py:410]: is the repetition of "id1" intentional? It doesn't seem to be relevant to the test execution.

File: tests/unit/test_extractor.py [hunk 4]
```diff
@@ -415,6 +405,33 @@ def test_upload_files_from_abfss_cognite_error(
     test_extractor.logger.error.assert_called_once()
 
 
+def test_retrieve_external_ids_from_lakehouse(test_extractor, mocker):
```
Comment (xtophs) [inline tests/unit/test_extractor.py:408]: Do we have a unit or integration test that asserts successful batching / filtering?

File: cdf_fabric_replicator/extractor.py (hunk 1 above)
Comment (muralidharm) [reply cdf_fabric_replicator/extractor.py:355]: Yes it should, good catch.  Since it seems to have worked with the incorrect parameter, I'm wondering if this is even needed for lakehouse reads.  Will investigate

File: tests/unit/test_extractor.py (hunk 4 above)
Comment (muralidharm) [reply tests/unit/test_extractor.py:408]: Yes, `test_extractor_timeseries_service_with_state` sets the state to a timestamp in the middle of the test data set and asserts the second half of the data was written to CDF>

File: tests/unit/test_extractor.py (hunk 3 above)
Comment (muralidharm) [reply tests/unit/test_extractor.py:410]: The goal of this test is to ensure that the method successfully returns unique values.  The 3rd "id1" might not be necessary, but I want to ensure duplicates are removed.

File: cdf_fabric_replicator/config.py (hunk 2 above)
Comment (muralidharm) [reply cdf_fabric_replicator/config.py:48]: Will add to documentation

File: cdf_fabric_replicator/extractor.py (hunk 1 above)
Comment (muralidharm) [reply cdf_fabric_replicator/extractor.py:355]: Interestingly `use_fabric_endpoint` seems to be needed only for write, not read.  The bearer token is still required.  Looking at those storage options they look like different ways to auth to a storage account - there might be something there for MSI but in past investigations I haven't been able to configure the MSI endpoint.  I would push that to another investigation, we've been using the guidance [here](https://blog.fabric.microsoft.com/en-us/blog/delta-rs-introduces-native-support-for-onelake/) to auth to Fabric.

File: cdf_fabric_replicator/extractor.py [hunk 5]
```diff
@@ -194,51 +195,61 @@ def upload_files_to_cdf(
...
-                                name=new_timeserie["externalId"],
-                                external_id=new_timeserie["externalId"],
-                                is_string=True,
-                                data_set_id=self.data_set_id,
-                            )
-                        )
-                except CogniteAPIError as e:
-                    self.logger.error(
-                        f"Error while writing time series data to CDF: {e}"
+            latest_timestamps[external_id] = self.state_store.get_state(state_id)[0]
```
Comment (muradsater) [inline cdf_fabric_replicator/extractor.py:219]: get_state can take a list of external ids as an argument, we should use that to avoid too many roundtrips if you have many external ids in your table
https://cognite-extractor-utils.readthedocs-hosted.com/en/latest/docs.html#cognite.extractorutils.statestore.AbstractStateStore.get_state

File: cdf_fabric_replicator/extractor.py [hunk 6]
```diff
@@ -304,19 +315,52 @@ def get_events(self, df: DataFrame) -> list[EventWrite]:
...
-    def convert_lakehouse_data_to_df(self, file_path: str, token: str) -> DataFrame:
+    def retrieve_external_ids_from_lakehouse(
+        self, file_path: str, token: str
+    ) -> list[str]:
         try:
-            dt = DeltaTable(
-                file_path,
-                storage_options={"bearer_token": token, "user_fabric_endpoint": "true"},
+            dt = DeltaTable(file_path, storage_options={"bearer_token": token})
+            table = dt.to_pyarrow_table(columns=["externalId"])
```
Comment (muradsater) [inline cdf_fabric_replicator/extractor.py:323]: Will this do the actual fetch of filtered data when the filter is applied?

File: cdf_fabric_replicator/extractor.py [hunk 7]
```diff
@@ -304,19 +315,52 @@ def get_events(self, df: DataFrame) -> list[EventWrite]:
...
             self.logger.error(
                 f"Error while converting lakehouse data to DataFrame: {e}"
             )
             raise e
 
+    def convert_lakehouse_data_to_df_batch(
+        self, file_path: str, external_id: str, timestamp: str, token: str
+    ) -> Generator[DataFrame, None, None]:
+        try:
+            dt = DeltaTable(file_path, storage_options={"bearer_token": token})
```
Comment (muradsater) [inline cdf_fabric_replicator/extractor.py:346]: Will this do the actual fetch of filtered data happen when the filter is applied is applied down at line 352?

File: cdf_fabric_replicator/extractor.py (hunk 5 above)
Comment (muralidharm) [reply cdf_fabric_replicator/extractor.py]: Thanks for the suggestion, modified code to use this

File: cdf_fabric_replicator/extractor.py [hunk 8]
```diff
@@ -194,51 +195,61 @@ def upload_files_to_cdf(
...
+        data_frame.index = pd.to_datetime(data_frame.index)
+        try:
+            self.client.time_series.data.insert_dataframe(data_frame)
+            self.logger.debug(f"Time series data written to CDF for {external_id}")
+            self.set_state(state_id, str(latest_process_time))
+        except CogniteNotFoundError as notFound:
+            missing_xids = notFound.not_found
+            for new_timeserie in missing_xids:
+                # Create the missing time series
+                self.client.time_series.create(
```
[the remark this answers is not in the document]
Comment (muralidharm) [reply cdf_fabric_replicator/extractor.py:244]: Added the insert dataframe call after creating time series.

File: cdf_fabric_replicator/extractor.py (hunk 6 above)
Comment (muralidharm) [reply cdf_fabric_replicator/extractor.py:332]: The data will be fetched on line 324 after the unique operation.  The data will also be restricted to the externalId column

File: tests/unit/test_extractor.py (hunk 3 above)
Comment (xtophs) [reply tests/unit/test_extractor.py]: it might be more clear if the 3rd value different from the first value "id1" to make abundantly clear that the 1st  item in the array was returned, not the 3rd..

**Review by muradsater — APPROVED** [review]
